### PR TITLE
Restrict Deployment of Matching Quote and Collateral tokens

### PR DESCRIPTION
**Pull request by ith-harvey**
# Description of change
## High level
* added a revert in the `canDeploy` pool modifier: `DeployQuoteCollateralSameToken()`.
* The revert fires if the `quote` and `collateral` addresses are equal to one another.

# Description of bug or vulnerability and solution
* Protocol allowed for pools that had the same `quote` and `collateral` token. 
  * Not a bug but an undesired user experience and could result in UI confusion / error

# Contract size
## Pre Change
```
  ERC721PoolFactory        -   3,340B  (13.59%)
  ERC20PoolFactory         -   2,473B  (10.06%)
```
## Post Change
```
  ERC721PoolFactory        -   3,390B  (13.79%)
  ERC20PoolFactory         -   2,523B  (10.27%)
```

# Gas usage
## Pre Change
```
| src/ERC20PoolFactory.sol:ERC20PoolFactory contract |                 |        |        |        |         |
|----------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                    | Deployment Size |        |        |        |         |
| 5393302                                            | 26753           |        |        |        |         |
| Function Name                                      | min             | avg    | median | max    | # calls |
| deployPool                                         | 602             | 226907 | 232241 | 255710 | 252     |
| deployedPools                                      | 854             | 2417   | 2854   | 2854   | 55      |
| deployedPoolsList                                  | 703             | 703    | 703    | 703    | 3       |

| src/ERC721PoolFactory.sol:ERC721PoolFactory contract |                 |                   |        |                     |         |
|------------------------------------------------------|-----------------|-------------------|--------|---------------------|---------|
| Deployment Cost                                      | Deployment Size |                   |        |                     |         |
| 5692525                                              | 28251           |                   |        |                     |         |
| Function Name                                        | min             | avg               | median | max                 | # calls |
| deployPool                                           | 926             | 66202914522633298 | 303974 | 8937393460516718734 | 135     |
| deployedPools                                        | 898             | 2461              | 2898   | 2898                | 55      |
| deployedPoolsList                                    | 614             | 614               | 614    | 614                 | 1       |
```
## Post Change
```
| src/ERC20PoolFactory.sol:ERC20PoolFactory contract |                 |        |        |        |         |
|----------------------------------------------------|-----------------|--------|--------|--------|---------|
| Deployment Cost                                    | Deployment Size |        |        |        |         |
| 5403305                                            | 26803           |        |        |        |         |
| Function Name                                      | min             | avg    | median | max    | # calls |
| deployPool                                         | 603             | 226105 | 232300 | 255769 | 254     |
| deployedPools                                      | 854             | 1970   | 2854   | 2854   | 77      |
| deployedPoolsList                                  | 703             | 703    | 703    | 703    | 3       |

| src/ERC721PoolFactory.sol:ERC721PoolFactory contract |                 |                   |        |                     |         |
|------------------------------------------------------|-----------------|-------------------|--------|---------------------|---------|
| Deployment Cost                                      | Deployment Size |                   |        |                     |         |
| 5702528                                              | 28301           |                   |        |                     |         |
| Function Name                                        | min             | avg               | median | max                 | # calls |
| deployPool                                           | 927             | 64297794680261978 | 304033 | 8937393460516718735 | 139     |
| deployedPools                                        | 898             | 2014              | 2898   | 2898                | 77      |
| deployedPoolsList                                    | 614             | 614               | 614    | 614                 | 1       |
```

### CODE REVIEW

File: tests/forge/unit/ERC721Pool/ERC721PoolFactory.t.sol [hunk 1]
```diff
@@ -115,15 +115,15 @@ contract ERC721PoolFactoryTest is ERC721HelperContract {
         _assertDeployWithInvalidRateRevert({
             poolFactory:  address(_factory),
             collateral:   address(_quote),
-            quote:        address(_quote),
+            quote:        address(new NFTCollateralToken()),
```
Comment (EdNoepel) [inline tests/forge/unit/ERC721Pool/ERC721PoolFactory.t.sol]: Shouldn't the NFT be the collateral token on the line above?

Comment (EdNoepel) [reply tests/forge/unit/ERC721Pool/ERC721PoolFactory.t.sol]: Same on lines 125-126.

**Review by EdNoepel — COMMENTED** [review]
Minor issue with `testDeployERC721CollectionPoolWithInvalidRate` test.  Factory changes look good.

File: tests/forge/unit/ERC721Pool/ERC721PoolFactory.t.sol (hunk 1 above)
Comment (ith-harvey) [reply tests/forge/unit/ERC721Pool/ERC721PoolFactory.t.sol]: Good catch, revised -> https://github.com/ajna-finance/contracts/pull/735/commits/b46b34b8f97dd1a79c935beb8d5ce70b8e0c7e9a

**Review by EdNoepel — APPROVED** [review]
👌

**Review by grandizzy — APPROVED** [review]

**Pull request merged**